utils.py

```diff
@@ -43,3 +62,31 @@ def default(self, obj):
       return obj.tolist()
     else:
       return super(NpEncoder, self).default(obj)
+
+
+def previous_matches_df(payroll, event):
+  # Reset index for faster deletions
+  payroll = payroll.set_index('Numéro Solde', drop=False)
+  # Remove duplicates from payroll
+  if 'payroll_duplicates' in event:
+    # Drop matched duplicates
+    for _, removed in event['payroll_duplicates']:
+      payroll = payroll.drop(removed)
+  # Convert matches to Series and left join them
+  duplicates = pd.Series(event['payroll_duplicates'], name="Duplicats de Payroll")
+  payroll = payroll.join(duplicates)
+  # Matched teachers
+  cols = ['', 'Matchs certains', 'Matchs confiants', 'Matchs possibles', 'Ghost teachers']
```

Comment:
Why the first col is `''`?

Reply:
Just because we start counting with `step1` and not `step0`

Reply:
It's just like doing `f"step{i-1}"` below

Reply:
That's what I suspected, ok 👌